### ISSUE TITLE
[How To] Passing JsonDocument to method of custom class which extracts JsonArray from it?

### ISSUE THREAD

**Issue by pranjal-joshi** [bblanchon/ArduinoJson#1289]
I am facing issue while passing a `DynamicJsonDocument` to the method of my class which is trying to extract JSON array and take some control actions accordingly.
```
**main sketch:**
SmartIO io;
io.setState(stateJson);

------------------------------------------------

**SmartIO: [MyClass]**
void SmartIO::setState(const JsonDocument doc) {
  byte i=0;
  JsonArray j = doc.as<JsonArray>();
  for(JsonVariant v : j) {
    if(v.as<byte>() == 1)
      // some control action here   
  }
}
```

With this code, I am getting following error:
```
sketch\SmartIO.cpp: In member function 'void SmartIo::setState(ArduinoJson6152_0000010::JsonDocument)':

SmartIO.cpp:60:35: error: conversion from 'ArduinoJson6152_0000010::VariantConstAs<ArduinoJson6152_0000010::ArrayRef>::type {aka ArduinoJson6152_0000010::ArrayConstRef}' to non-scalar type 'ArduinoJson::JsonArray {aka ArduinoJson6152_0000010::ArrayRef}' requested

   JsonArray j = doc.as<JsonArray>();

                                   ^

exit status 1
In file included from C:\Users\Pranjal Joshi\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson/Document/BasicJsonDocument.hpp:7:0,

                 from C:\Users\Pranjal Joshi\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson/Document/DynamicJsonDocument.hpp:7,

                 from C:\Users\Pranjal Joshi\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson.hpp:21,

                 from C:\Users\Pranjal Joshi\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson.h:9,

                 from C:\Users\Pranjal Joshi\Documents\Arduino\libraries\Painless_Mesh\src/painlessmesh/configuration.hpp:14,

                 from C:\Users\Pranjal Joshi\Documents\Arduino\libraries\Painless_Mesh\src/painlessMesh.h:11,

                 from E:\SMART\v3\ESP\smart_v3\smart_v3.ino:8:

C:\Users\Pranjal Joshi\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson/Document/JsonDocument.hpp: In function 'void setup()':

C:\Users\Pranjal Joshi\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson/Document/JsonDocument.hpp:324:3: error: 'ArduinoJson6152_1000010::JsonDocument::JsonDocument(const ArduinoJson6152_1000010::JsonDocument&)' is private

   JsonDocument(const JsonDocument&);

   ^

smart_v3:69:26: error: within this context

     io.setState(stateJson);

                          ^

In file included from E:\SMART\v3\ESP\smart_v3\smart_v3.ino:15:0:

SmartIO.h:21:10: error: initializing argument 1 of 'void SmartIo::setState(ArduinoJson6152_1000010::JsonDocument)'

     void setState(const JsonDocument doc);

          ^

conversion from 'ArduinoJson6152_0000010::VariantConstAs<ArduinoJson6152_0000010::ArrayRef>::type {aka ArduinoJson6152_0000010::ArrayConstRef}' to non-scalar type 'ArduinoJson::JsonArray {aka ArduinoJson6152_0000010::ArrayRef}' requested

```

**I have tried various methods like passing by reference, passing pointers.. but as of now, nothing has compiled successfully!**

**Comment by ewaldc** [comment]
I think JsonDocument is a private reference,
You can use *JsonDocument or `void setstate (JsonVariant doc)` with `setstate(doc.as<JsonVariant>());`

If you want it immutable, you can use _JsonVariantConst_

**Comment by pranjal-joshi** [comment]
@ewaldc After changing the argument data type to `JsonVariant`, I got following type conversion error!
```
sketch\SmartIO.cpp: In member function 'void SmartIo::setState(ArduinoJson::JsonVariantConst)':

SmartIO.cpp:60:35: error: conversion from 'ArduinoJson6152_0000010::VariantConstAs<ArduinoJson6152_0000010::ArrayRef>::type {aka ArduinoJson6152_0000010::ArrayConstRef}' to non-scalar type 'ArduinoJson::JsonArray {aka ArduinoJson6152_0000010::ArrayRef}' requested

   JsonArray j = doc.as<JsonArray>();

                                   ^

exit status 1
conversion from 'ArduinoJson6152_0000010::VariantConstAs<ArduinoJson6152_0000010::ArrayRef>::type {aka ArduinoJson6152_0000010::ArrayConstRef}' to non-scalar type 'ArduinoJson::JsonArray {aka ArduinoJson6152_0000010::ArrayRef}' requested

```

**UPDATE:**
If I change argument from `JsonVariantConst` to `JsonVariant` and pass `io.setState(stateJson.as<JsonVariant>())`, I get the following linker error:
```
c:/users/pranjal/appdata/local/arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/2.5.0-4-b40a506/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: sketch\smart_v3.ino.cpp.o:C:\Users\Pranjal\Documents\Arduino\libraries\Painless_Mesh\src/painlessmesh/mesh.hpp:216: undefined reference to `SmartIo::setState(ArduinoJson6152_1000010::VariantRef)'

c:/users/pranjal/appdata/local/arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/2.5.0-4-b40a506/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: sketch\smart_v3.ino.cpp.o:c:\users\pranjal\appdata\local\arduino15\packages\esp8266\tools\xtensa-lx106-elf-gcc\2.5.0-4-b40a506\xtensa-lx106-elf\include\c++\4.8.2/functional:2453: undefined reference to `SmartIo::setState(ArduinoJson6152_1000010::VariantRef)'

collect2.exe: error: ld returned 1 exit status

exit status 1
```

**Comment by ewaldc** [comment]
Sorry, I am not seeing the problem...
```
void setstate (JsonVariant doc) {
  JsonArray j = doc.as<JsonArray>();
  for (const auto& v : j) {
    Serial.print("Value: - ");
    Serial.println(v.as<String>());
    Serial.println(v.as<const float>());
      // some control action here   
  }
}
DynamicJsonDocument doc(200);
char json[] = "{\"sensor\":\"gps\",\"time\":1351824120,\"data\":[48.756080,2.302038]}";
DeserializationError error = deserializeJson(doc, json);
if (!error) {
  JsonVariant varDoc = doc.as<JsonObject>();
  JsonVariant varArr = doc["data"];
  setstate(varArr);
}
```
If _json_ is an array, you can pass _doc.as<JsonVariant>()_ but you can't pass a JsonObject and then process it as a JsonArray.

**Comment by bblanchon** [comment]
Hi @pranjal-joshi,

> ```
> error: 'JsonDocument::JsonDocument(const JsonDocument&)' is private
> ```

You must pass [`JsonDocument`](https://arduinojson.org/v6/api/jsondocument/) by reference because the copy constructor of [`JsonDocument`](https://arduinojson.org/v6/api/jsondocument/) is private. If you do need to make a copy (which doesn't seem to be the case, here), use either [`DynamicJsonDocument`](https://arduinojson.org/v6/api/dynamicjsondocument/) or [`StaticJsonDocument`](https://arduinojson.org/v6/api/staticjsondocument/).

> ```
> error: conversion from 'ArrayConstRef' to non-scalar type 'ArrayRef' requested
> ```

You cannot get a [`JsonArray`](https://arduinojson.org/v6/api/jsonarray/) pointing to a `const JsonDocument` because it would allow you to mutate the read-only document. That's why `const JsonDocument` returns a `ArrayConstRef`, a.k.a. [`JsonArrayConst`](https://arduinojson.org/v6/api/jsonarrayconst/), when you ask for [`JsonArray`](https://arduinojson.org/v6/api/jsonarray/).

To summarize, here is how you can fix your program:

```c++
void SmartIO::setState(const JsonDocument& doc) {
  JsonArrayConst j = doc.as<JsonArrayConst>();
  for (JsonVariantConst v : j) {
    if (v == 1)
      // some control action here   
  }
}
```

Notice that I also changed [`JsonVariant`](https://arduinojson.org/v6/api/jsonvariant/) to [`JsonVariantConst`](https://arduinojson.org/v6/api/jsonvariantconst/) for the same reason.

Best regards,
Benoit

**Comment by pranjal-joshi** [comment]
@bblanchon 
Hello, After trying to implement using `JsonDocumnet&` passing by reference throws the following error. Note that I have tried both with and without `const` datatypes but it throws the same error!
```
c:/users/pranjal/appdata/local/arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/2.5.0-4-b40a506/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: sketch\smart_v3.ino.cpp.o:C:\Users\Pranjal\Documents\Arduino\libraries\ArduinoJson\src/ArduinoJson/Json/JsonSerializer.hpp:82: undefined reference to `SmartIo::setState(ArduinoJson6152_1000010::JsonDocument const&)'

c:/users/pranjal/appdata/local/arduino15/packages/esp8266/tools/xtensa-lx106-elf-gcc/2.5.0-4-b40a506/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: sketch\smart_v3.ino.cpp.o: in function `looper()':

F:\Personal\Code\SMART\v3\ESP\smart_v3/smart_v3.ino:130: undefined reference to `SmartIo::setState(ArduinoJson6152_1000010::JsonDocument const&)'

collect2.exe: error: ld returned 1 exit status

exit status 1
```

**Sugeestion:** It would be great if you **add examples of passing and returning data** similar to the serialization and deserialization of all JsonVariants in the documentation.

Currently, I am getting the work done by passing data between methods by `deserializeJson` to the `char buf` and then `serializeJson` inside the method but it's kinda dirty to do that!

**Comment by bblanchon** [comment]
Hi @pranjal-joshi,

I believe you made a mistake while implementing the solution I propose.
I cannot tell what the problem is without looking at the code.

I'd be happy to add examples according to your suggestion, but I would first need to understand your use case. Please explain what you're trying to do.

Best regards,
Benoit

**Comment by bblanchon** [comment]
Any progress, @pranjal-joshi?

**Comment by pranjal-joshi** [comment]
> Any progress, @pranjal-joshi?

Nope! Still using same workaround to get things done.
`Serialize to char buf -> pass  -> deserialize it in class method`
